File: library/src/main/java/org/meganekkovr/SurfaceRendererComponent.java
package org.meganekkovr;

import android.graphics.Canvas;
import android.graphics.PorterDuff;
import android.graphics.SurfaceTexture;
import android.graphics.drawable.Drawable;
import android.view.Surface;
import android.view.View;
import android.view.ViewGroup;

/**
 * This gives appearance to {@link Entity} for rendering.
 * Surface can be defined with {@link Drawable}, {@link View}, or {@link android.graphics.Bitmap}.
 * You can use {@link Canvas} to render custom drawings.
 * {@link #getSurfaceTexture()} or {@link #getSurface()} are usefull if you wish to render video
 * or camera images with {@link android.media.MediaPlayer} or {@link android.hardware.Camera}.
 */
public class SurfaceRendererComponent extends Component {

    private final NativePointer nativePointer;
    private CanvasRenderer canvasRenderer;
    private boolean continuousUpdate;
    private float opacity = 1.0f;
    private StereoMode stereoMode = StereoMode.NORMAL;
    public SurfaceRendererComponent() {
        nativePointer = NativePointer.getInstance(newInstance());
    }

    private static native SurfaceTexture getSurfaceTexture(long nativePtr);

    private static native Surface getSurface(long nativePtr);

    private static native void setEntityTexture(long entityPtr, long nativePtr);

    private static native void removeEntityTexture(long entityPtr, long nativePtr);

    private static native void setOpacity(long nativePtr, float opacity);

    private static native void setStereoMode(long nativePtr, int stereoMode);

    private static native void setUseChromaKey(long nativePtr, boolean useChromaKey);

    private static native boolean getUseChromaKey(long nativePtr);

    private static native void setChromaKeyColor(long nativePtr, float r, float g, float b);

    public static SurfaceRendererComponent from(View view) {

        view.measure(0, 0);
        final int width = view.getMeasuredWidth();
        final int height = view.getMeasuredHeight();
        view.layout(0, 0, width, height);

        SurfaceRendererComponent component = new SurfaceRendererComponent();
        component.setCanvasRenderer(ViewRenderer.from(view));

        return component;
    }

    public static SurfaceRendererComponent from(Drawable drawable) {

        drawable.setBounds(0, 0, drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());

        SurfaceRendererComponent component = new SurfaceRendererComponent();
        component.setCanvasRenderer(new DrawableRenderer(drawable));

        return component;
    }

    protected native long newInstance();

    @Override
    public void onAttach(Entity entity) {
        super.onAttach(entity);
        setEntityTexture(entity.getNativePointer(), nativePointer.get());
    }

    @Override
    public void onDetach(Entity entity) {
        super.onDetach(entity);
        removeEntityTexture(entity.getNativePointer(), nativePointer.get());
    }

    @Override
    public void update(FrameInput frame) {

        if (canvasRenderer != null) {
            if (canvasRenderer.isDirty()) {
                Surface surface = getSurface(nativePointer.get());
                Canvas canvas = surface.lockCanvas(null);

                // One time draw if return true
                if (canvasRenderer.render(canvas)) {
                    canvasRenderer.dirty = false;
                }

                surface.unlockCanvasAndPost(canvas);

                // One time update texture
                if (!continuousUpdate) getSurfaceTexture().updateTexImage();
            }
        }

        // Continuous update texture
        if (continuousUpdate) {
            getSurfaceTexture().updateTexImage();
        }

        super.update(frame);
    }

    public SurfaceTexture getSurfaceTexture() {
        return getSurfaceTexture(nativePointer.get());
    }

    public Surface getSurface() {
        return getSurface(nativePointer.get());
    }

    public CanvasRenderer getCanvasRenderer() {
        return canvasRenderer;
    }

    public void setCanvasRenderer(CanvasRenderer canvasRenderer) {
        this.canvasRenderer = canvasRenderer;

        if (canvasRenderer != null) {
            getSurfaceTexture().setDefaultBufferSize(canvasRenderer.width, canvasRenderer.height);
        }
    }

    /**
     * Call this to update internal {@link SurfaceTexture} on every frame update.
     * Typically used when rendering video with {@link android.media.MediaPlayer}.
     *
     * @param continuousUpdate {@code true} to update texture automatically on frame update. default is {@code false}.
     */
    public void setContinuousUpdate(boolean continuousUpdate) {
        this.continuousUpdate = continuousUpdate;
    }

    /**
     * Get actual opacity for rendering. This value is inherited from parent {@link Entity}.
     *
     * @return Actual opacity for rendering
     */
    public float getOpacity() {
        return opacity;
    }

    void setOpacity(float opacity) {

        if (opacity < 0) {
            opacity = 0;
        } else if (opacity > 1) {
            opacity = 1;
        }

        this.opacity = opacity;
        setOpacity(nativePointer.get(), opacity);
    }

    public StereoMode getStereoMode() {
        return stereoMode;
    }

    public void setStereoMode(StereoMode stereoMode) {
        this.stereoMode = stereoMode;
        setStereoMode(nativePointer.get(), stereoMode.ordinal());
    }

    public boolean getUseChromaKey() {
        return getUseChromaKey(nativePointer.get());
    }

    public void setUseChromaKey(boolean useChromaKey) {
        setUseChromaKey(nativePointer.get(), useChromaKey);
    }

    public void setChromaKeyColor(float r, float g, float b) {
        setChromaKeyColor(nativePointer.get(), r, g, b);
    }

    public enum StereoMode {
        NORMAL,      // 0
        TOP_BOTTOM,  // 1
        BOTTOM_TOP,  // 2
        LEFT_RIGHT,  // 3
        RIGHT_LEFT,  // 4
        TOP_ONLY,    // 5
        BOTTOM_ONLY, // 6
        LEFT_ONLY,   // 7
        RIGHT_ONLY   // 8
    }

    public static abstract class CanvasRenderer {
        public final int width;
        public final int height;
        private boolean dirty = true;

        public CanvasRenderer(int width, int height) {
            this.width = width;
            this.height = height;
        }

        public boolean isDirty() {
            return dirty;
        }

        /**
         * @param canvas Canvas to draw.
         * @return {@code true} to end drawing. No more called in next update. If redrawing is required, call {@link #invalidate()}.
         * {@code false} to continues draw. This behavior can be modified overriding {@link #isDirty()}.
         */
        protected abstract boolean render(Canvas canvas);

        public void invalidate() {
            dirty = true;
        }
    }

    public static class DrawableRenderer extends CanvasRenderer {
        private final Drawable drawable;

        public DrawableRenderer(Drawable drawable) {
            super(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());
            this.drawable = drawable;
        }

        @Override
        protected boolean render(Canvas canvas) {
            drawable.draw(canvas);
            return true;
        }
    }

    public static class ViewRenderer extends CanvasRenderer {
        private final View view;

        protected ViewRenderer(View view, int width, int height) {
            super(width, height);
            this.view = view;
        }

        public static ViewRenderer from(View view) {
            view.measure(0, 0);
            final int width = view.getMeasuredWidth();
            final int height = view.getMeasuredHeight();
            view.layout(0, 0, width, height);
            return new ViewRenderer(view, width, height);
        }

        /**
         * Check dirty state of view recursively.
         *
         * @param view Checked View.
         * @return Returns true if at least one View is dirty in hierarchy.
         */
        private static boolean isDirty(View view) {

            if (view.isDirty()) return true;

            // Apply this method to all children of view if view is ViewGroup
            if (view instanceof ViewGroup) {
                final ViewGroup viewGroup = (ViewGroup) view;

                for (int i = 0, count = viewGroup.getChildCount(); i < count; ++i) {
                    if (isDirty(viewGroup.getChildAt(i))) return true;
                }
            }

            return false;
        }

        @Override
        public boolean isDirty() {
            return isDirty(view);
        }

        @Override
        protected boolean render(Canvas canvas) {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR);
            view.draw(canvas);
            return true;
        }

        public View getView() {
            return view;
        }
    }
}

File: library/src/main/java/org/meganekkovr/SurfaceRendererComponent.kt
package org.meganekkovr

import android.graphics.Canvas
import android.graphics.PorterDuff
import android.graphics.SurfaceTexture
import android.graphics.drawable.Drawable
import android.view.Surface
import android.view.View
import android.view.ViewGroup

/**
 * This gives appearance to [Entity] for rendering.
 * Surface can be defined with [Drawable], [View], or [android.graphics.Bitmap].
 * You can use [Canvas] to render custom drawings.
 * [.getSurfaceTexture] or [.getSurface] are usefull if you wish to render video
 * or camera images with [android.media.MediaPlayer] or [android.hardware.Camera].
 */
class SurfaceRendererComponent : Component() {

    private val nativePointer = NativePointer.getInstance(newInstance())
    var canvasRenderer: CanvasRenderer? = null
        set(canvasRenderer) {
            field = canvasRenderer

            if (canvasRenderer != null) {
                surfaceTexture.setDefaultBufferSize(canvasRenderer.width, canvasRenderer.height)
            }
        }

    var continuousUpdate: Boolean = false

    /**
     * Get actual opacity for rendering. This value is inherited from parent [Entity].
     *
     * @return Actual opacity for rendering
     */
    var opacity = 1.0f
        set(opacity) {

            field = if (opacity < 0) {
                0f
            } else if (opacity > 1) {
                1f
            } else {
                opacity
            }

            setOpacity(nativePointer.get(), opacity)
        }

    var stereoMode = StereoMode.NORMAL
        set(stereoMode) {
            field = stereoMode
            setStereoMode(nativePointer.get(), stereoMode.ordinal)
        }

    val surfaceTexture: SurfaceTexture
        get() = getSurfaceTexture(nativePointer.get())

    val surface: Surface
        get() = getSurface(nativePointer.get())

    var useChromaKey: Boolean
        get() = getUseChromaKey(nativePointer.get())
        set(useChromaKey) = setUseChromaKey(nativePointer.get(), useChromaKey)

    var chromaKeyThreshold: Float
        get() = getChromaKeyThreshold(nativePointer.get())
        set(chromaKeyThreshold) = setChromaKeyThreshold(nativePointer.get(), chromaKeyThreshold)

    var chromaKeyBlend: Float
        get() = getChromaKeyBlend(nativePointer.get())
        set(chromaKeyBlend) = setChromaKeyBlend(nativePointer.get(), chromaKeyBlend)

    private external fun getSurfaceTexture(nativePtr: Long): SurfaceTexture

    private external fun getSurface(nativePtr: Long): Surface

    private external fun setEntityTexture(entityPtr: Long, nativePtr: Long)

    private external fun removeEntityTexture(entityPtr: Long, nativePtr: Long)

    private external fun setOpacity(nativePtr: Long, opacity: Float)

    private external fun setStereoMode(nativePtr: Long, stereoMode: Int)

    private external fun setUseChromaKey(nativePtr: Long, useChromaKey: Boolean)

    private external fun getUseChromaKey(nativePtr: Long): Boolean

    private external fun setChromaKeyThreshold(nativePtr: Long, chromaKeyThreshold: Float)

    private external fun getChromaKeyThreshold(nativePtr: Long): Float

    private external fun setChromaKeyBlend(nativePtr: Long, chromaKeyBlend: Float)

    private external fun getChromaKeyBlend(nativePtr: Long): Float

    private external fun setChromaKeyColor(nativePtr: Long, r: Float, g: Float, b: Float)

    private external fun newInstance(): Long

    override fun onAttach(entity: Entity) {
        super.onAttach(entity)
        setEntityTexture(entity.nativePointer, nativePointer.get())
    }

    override fun onDetach(entity: Entity) {
        super.onDetach(entity)
        removeEntityTexture(entity.nativePointer, nativePointer.get())
    }

    override fun update(frame: FrameInput) {

        val renderer = this.canvasRenderer
        if (renderer != null) {

            if (renderer.isDirty) {
                val surface = getSurface(nativePointer.get())
                val canvas = surface.lockCanvas(null)

                // One time draw if return true
                if (renderer.render(canvas)) {
                    renderer.isDirty = false
                }

                surface.unlockCanvasAndPost(canvas)

                // One time update texture
                if (!continuousUpdate) surfaceTexture.updateTexImage()
            }
        }

        // Continuous update texture
        if (continuousUpdate) {
            surfaceTexture.updateTexImage()
        }

        super.update(frame)
    }

    fun setChromaKeyColor(r: Float, g: Float, b: Float) {
        setChromaKeyColor(nativePointer.get(), r, g, b)
    }

    enum class StereoMode {
        NORMAL, // 0
        TOP_BOTTOM, // 1
        BOTTOM_TOP, // 2
        LEFT_RIGHT, // 3
        RIGHT_LEFT, // 4
        TOP_ONLY, // 5
        BOTTOM_ONLY, // 6
        LEFT_ONLY, // 7
        RIGHT_ONLY   // 8
    }

    abstract class CanvasRenderer(val width: Int, val height: Int) {
        open var isDirty = true

        /**
         * @param canvas Canvas to draw.
         * @return `true` to end drawing. No more called in next update. If redrawing is required, call [.invalidate].
         * `false` to continues draw. This behavior can be modified overriding [.isDirty].
         */
        abstract fun render(canvas: Canvas): Boolean

        fun invalidate() {
            isDirty = true
        }
    }

    class DrawableRenderer(private val drawable: Drawable) : CanvasRenderer(drawable.intrinsicWidth, drawable.intrinsicHeight) {

        override fun render(canvas: Canvas): Boolean {
            drawable.draw(canvas)
            return true
        }
    }

    class ViewRenderer protected constructor(val view: View, width: Int, height: Int) : CanvasRenderer(width, height) {

        override var isDirty: Boolean
            get() = isDirty(view)
            set(value) {
                super.isDirty = value
            }

        override fun render(canvas: Canvas): Boolean {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR)
            view.draw(canvas)
            return true
        }

        companion object {

            fun from(view: View): ViewRenderer {
                view.measure(0, 0)
                val width = view.measuredWidth
                val height = view.measuredHeight
                view.layout(0, 0, width, height)
                return ViewRenderer(view, width, height)
            }

            /**
             * Check dirty state of view recursively.
             *
             * @param view Checked View.
             * @return Returns true if at least one View is dirty in hierarchy.
             */
            private fun isDirty(view: View): Boolean {

                if (view.isDirty) return true

                // Apply this method to all children of view if view is ViewGroup
                if (view is ViewGroup) {

                    var i = 0
                    val count = view.childCount
                    while (i < count) {
                        if (isDirty(view.getChildAt(i))) return true
                        ++i
                    }
                }

                return false
            }
        }
    }

    companion object {

        @JvmStatic
        fun from(view: View): SurfaceRendererComponent {

            view.measure(0, 0)
            val width = view.measuredWidth
            val height = view.measuredHeight
            view.layout(0, 0, width, height)

            val component = SurfaceRendererComponent()
            component.canvasRenderer = ViewRenderer.from(view)

            return component
        }

        @JvmStatic
        fun from(drawable: Drawable): SurfaceRendererComponent {

            drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)

            val component = SurfaceRendererComponent()
            component.canvasRenderer = DrawableRenderer(drawable)

            return component
        }
    }
}
